### PR TITLE
Add a 'reservation expiration' mechanism

### PR DESCRIPTION
**Pull request by niftynei**
It's useful to be able to 'retire' reservations on utxos after a certain time lapse. It's also useful to be able to RBF transactions. This PR lets us do both now, by adding two new parameters to `reserveinputs`. 

- `expire_after` which will functionally 'unreserve' a utxo after that many blocks have passed, and
- `allow_rbf` which if set will create a psbt that has all of its inputs set to be rbf'able. 

RBF'ability forced a few not so nice state breaks in the output status (you have to be able to select utxos that are marked as 'spent' but not in a block yet, etc). But hey, RBF, here it is!

Built ontop of the as of yet unmerged #3775 (new commits start at https://github.com/ElementsProject/lightning/commit/37ba229563002e2962e0d18348c2d6509f1895fa)

### CODE REVIEW

File: wallet/walletrpc.c [hunk 1]
```diff
@@ -1210,9 +1210,9 @@ static struct command_result *json_reserveinputs(struct command *cmd,
 	 * around, so we remove the auto-cleanup that happens
 	 * when the utxo objects are free'd */
 	wallet_persist_utxo_reservation(cmd->ld->wallet, utx->wtx->utxos);
-	/* Set the lease on these utxos to expire in ~4 hours,
+	/* Set the lease on these utxos to expire in ~24 hours,
 	 * or 24 blocks */
-	u32 expires_at = cmd->ld->topology->tip->height + 6 * 4;
+	u32 expires_at = cmd->ld->topology->tip->height + 6 * 24;
```
Comment (cdecker) [inline wallet/walletrpc.c]: The commit message needs an update if we change this default value.

File: wallet/walletrpc.c [hunk 2]
```diff
@@ -160,6 +160,147 @@ static struct command_result *broadcast_and_wait(struct command *cmd,
...
+	if (result)
+		return result;
+
+	/* Because of the max limit of AMOUNT_SAT(-1ULL),
+	 * `(*utx)->wtx->all_funds` won't change in `wtx_select_utxos()` */
+	if ((*utx)->wtx->all_funds)
+		outputs[0]->amount = (*utx)->wtx->amount;
+
+	/* Add the change as the last output */
+	if (!amount_sat_eq((*utx)->wtx->change, AMOUNT_SAT(0))) {
```
Comment (cdecker) [inline wallet/walletrpc.c:271]: Does wally not set `wtx->change` if it is below dust? If that's not the case we should probably compare against the dust limit in `chainparams`.

File: wallet/walletrpc.c [hunk 3]
```diff
@@ -160,6 +160,147 @@ static struct command_result *broadcast_and_wait(struct command *cmd,
...
+		struct bitcoin_tx_output *change_output;
+
+		changekey = tal(tmpctx, struct pubkey);
+		if (!bip32_pubkey(cmd->ld->wallet->bip32_base, changekey,
+				  (*utx)->wtx->change_key_index))
+			return command_fail(cmd, LIGHTNINGD, "Keys generation failure");
```
Comment (cdecker) [inline wallet/walletrpc.c:277]:
```suggestion
			return command_fail(cmd, LIGHTNINGD, "Change key generation failed");
```

File: wallet/walletrpc.c [hunk 4]
```diff
@@ -1243,14 +1243,15 @@ static struct command_result *json_reserveinputs(struct command *cmd,
 	struct unreleased_tx *utx;
 	const struct utxo **chosen_utxos;
 	struct bitcoin_tx_output **outputs;
-	u32 *feerate_per_kw, *minconf;
+	u32 *feerate_per_kw, *minconf, *expire_at;
```
Comment (cdecker) [inline wallet/walletrpc.c:1248]: Just to match the param name:

```suggestion
	u32 *feerate_per_kw, *minconf, *expire_after;
```

File: wallet/walletrpc.c [hunk 5]
```diff
@@ -1275,13 +1276,12 @@ static struct command_result *json_reserveinputs(struct command *cmd,
...
-	u32 expires_at = cmd->ld->topology->tip->height + 6 * 24;
-	for (size_t i = 0; i < tal_count(utx->wtx->utxos); i++)
-		wallet_output_reservation_update(cmd->ld->wallet,
-						 utx->wtx->utxos[i],
-						 expires_at);
+	if (*expire_at > 0) {
+		for (size_t i = 0; i < tal_count(utx->wtx->utxos); i++)
+			wallet_output_reservation_update(cmd->ld->wallet,
+							 utx->wtx->utxos[i],
+							 *expire_at);
```
Comment (cdecker) [inline wallet/walletrpc.c:1287]: Seems to me like this uses the relative expiry as absolute blockheight. Should probably be something like the following:

```suggestion
		for (size_t i = 0; i < tal_count(utx->wtx->utxos); i++)
			wallet_output_reservation_update(cmd->ld->wallet,
							 utx->wtx->utxos[i],
							 cmd->ld->topology->tip->height + *expire_at);
```

File: common/withdraw_tx.c [hunk 6]
```diff
@@ -12,17 +12,29 @@
...
+	 * BIP-125: Opt-in Full Replace-by-Fee Signaling
+	 *     https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki
+	 * A transaction is considered to have opted in to
+	 * allowing replacement of itself if any of its
+	 * inputs have an nSequence number less than (0xffffffff - 1).
+	 */
+	/* A sequence of (0xffffffff - 1) signals to use locktime */
+	u32 sequence = BITCOIN_TX_DEFAULT_SEQUENCE - 1;
+	if (allow_rbf)
+		sequence--;
```
Comment (cdecker) [inline common/withdraw_tx.c:33]: Isn't this now `0xffffffff - 2` because `BITCOIN_TX_DEFAULT_SEQUENCE` is `0xffffffff` and we subtract twice? That'd mean we always opt in.

File: common/wallet_tx.c [hunk 7]
```diff
@@ -91,11 +91,6 @@ struct command_result *param_utxos(struct command *cmd,
 	tal_free(txids);
 	tal_free(outnums);
 
-	if (!*utxos)
-		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
-							"Could not decode all of the outpoints. The utxos"
-							" should be specified as an array of "
-							" 'txid:output_index'.");
```
Comment (cdecker) [inline common/wallet_tx.c:98]: This looks like an unrelated change.

File: tests/test_wallet.py [hunk 8]
```diff
@@ -586,6 +586,59 @@ def test_reserveinputs(node_factory, bitcoind, chainparams):
...
+    psbt = bitcoind.rpc.decodepsbt(reserved['psbt'])
+    utxos = []
+    unreserve_utxos = []
+    for vin in psbt['tx']['vin']:
+        utxos.append(vin['txid'] + ":" + str(vin['vout']))
+        unreserve_utxos.append({'txid': vin['txid'], 'vout': vin['vout'], 'sequence': vin['sequence']})
+    unreserve_psbt = bitcoind.rpc.createpsbt(unreserve_utxos, [])
+    unres = l1.rpc.unreserveinputs(unreserve_psbt)
+
+    # let's not set the fee high enough, should explodes
```
Comment (cdecker) [inline tests/test_wallet.py:628]:
```suggestion
    # let's not set the fee high enough, should explode
```

File: tests/test_wallet.py [hunk 9]
```diff
@@ -586,6 +586,59 @@ def test_reserveinputs(node_factory, bitcoind, chainparams):
...
+    signed_psbt = l1.rpc.signpsbt(reserved['psbt'])['signed_psbt']
+    with pytest.raises(RpcError, match=r'insufficient fee, rejecting replacement'):
+        l1.rpc.sendpsbt(signed_psbt)
+
+    # now let's do it for reals
+    unres = l1.rpc.unreserveinputs(unreserve_psbt)
+    reserved = l1.rpc.reserveinputs([{addr: Millisatoshi(amount * 0.5 * 1000)}], feerate='10000perkw', utxos=utxos)
+    signed_psbt = l1.rpc.signpsbt(reserved['psbt'])['signed_psbt']
+    l1.rpc.sendpsbt(signed_psbt)
+    assert bitcoind.rpc.getmempoolinfo()['size'] == 1
```
Comment (cdecker) [inline tests/test_wallet.py:639]: Should we check that the new tx is in the mempool? Just checking for the size of the mempool might pass despite the old tx still being in there. `bitcoind.rpc.getrawmempool()` will give you the txids.

**Comment by rustyrussell** [comment]
Note: after discussing with @niftynei I agree with this concept, but she didn't take it far enough.

Because we didn't expire utxo reservations, we had multiple kludges:
1. On startup, we unreserve all of them.
2. We reserve them, but use a destructor to unreserve them.
3. We mark them *spent* before we see the transaction in the blockchain, leading to missing funds and rescan if it never confirms.

On digging further, I noticed that with a simple change, we can move utxo.h entirely into the wallet dir. This further allows us to simplify our utxo handling.

So I've cherry picked part of the "reservations expire" commit and started ripping things apart...

File: wallet/wallet.c [hunk 10]
```diff
@@ -229,6 +245,79 @@ bool wallet_update_output_status(struct wallet *w,
...
+	stmt = db_prepare_v2(w->db, SQL("SELECT COUNT(*) FROM outputs "
+				"WHERE status=? AND prev_out_tx=? "
+				"AND prev_out_index=?"));
+	db_bind_int(stmt, 0, output_status_in_db(state));
+	db_bind_txid(stmt, 1, txid);
+	db_bind_int(stmt, 2, outnum);
+
+	db_query_prepared(stmt);
+	if (db_step(stmt)) {
+		res = db_column_u64(stmt, 0) > 0;
```
[the remark this answers is not in the document]
Comment (niftynei) [reply wallet/wallet.c:265]: per their [docs](https://www.postgresql.org/docs/current/functions-aggregate.html), it returns a `bigint` which is a signed 8-byte int

File: wallet/wallet.c [hunk 11]
```diff
@@ -355,10 +448,89 @@ struct utxo *wallet_utxo_get(const tal_t *ctx, struct wallet *w,
...
+	struct utxo **reserved, **available;
+	size_t rcount;
+
+	reserved = expired_reserved_utxos_get(ctx, w);
+	available = wallet_get_utxos(ctx, w, output_state_available);
+
+	rcount = tal_count(reserved);
+	tal_resize(&reserved, rcount + tal_count(available));
+	for (size_t i = 0; i < tal_count(available); i++)
+		reserved[rcount + i] = tal_steal(reserved, available[i]);
```
[the remark this answers is not in the document]
Comment (niftynei) [reply wallet/wallet.c:504]: yes but then we'd have to clean them up on every new block. Another option would be to remove the 'output_state_available'/ status altogether and just use the reserved_til as a marker of whether or not the utxo is still reserved.

File: common/withdraw_tx.c (hunk 6 above)
Comment (niftynei) [reply common/withdraw_tx.c:33]: Right, the intention is to opt in to RBF if 'allow_rbf' is true.

`0xffffffff` is no rbf, no locktime
`0xffffffff - 1` is no rbf, allow locktime (which we set on all withdraws, and now fundchannels)
`0xffffffff - 2` is allow rbf, allow locktime

File: wallet/walletrpc.c (hunk 5 above)
Comment (niftynei) [reply wallet/walletrpc.c:1287]: Ah yes, good spot.

File: wallet/walletrpc.c (hunk 2 above)
Comment (niftynei) [reply wallet/walletrpc.c:271]: `wtx_select_utxos` does the dust comparison and sets this to zero if it's less than that, so this is an accurate check.

**Comment by niftynei** [comment]
This will largely be subsumed by work @rustyrussell is doing on the utxo infrastructure, but updated with comments for good measure.

**Comment by cdecker** [comment]
> This will largely be subsumed by work @rustyrussell is doing on the utxo infrastructure, but updated with comments for good measure.

Thanks for the heads up :+1:

**Comment by ZmnSCPxj** [comment]
>> This will largely be subsumed by work @rustyrussell is doing on the utxo infrastructure, but updated with comments for good measure.
>
>Thanks for the heads up :+1:

I thought this was #3798 , which was withdrawn?  The nearest equivalent is #3821, which does not look like it has reservation expiration.  What is the status?  How does this affect #3763 which apparently need to be blocked for the new UTXO things?
